keps/sig-network/3453-minimize-iptables-restore/README.md

```diff
@@ -0,0 +1,473 @@
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The performance of kube-proxy's iptables mode in large clusters could
+be greatly increased by optimizing its calls to `iptables-restore`.
+Although this is a small change in terms of code size, it has a large
+effect on kube-proxy operation, and could potentially completely break
+clusters (if the code was buggy), so we have decided to treat it as a
+KEP-able feature.
+
+## Motivation
+
+### Goals
+
+- Improve the performance of kube-proxy in iptables mode, by optimizing
+  its calls to `iptables-restore` by not including rules that haven't
+  changed.
+
+### Non-Goals
+
+- Everything else.
+
+## Proposal
+
+### General Plan
+
+`iptables-restore` has the constraint that if you want to specify
+_any_ rule in a given iptables chain, then you must specify _every_
+rule in that chain. However, if you want to leave a chain completely
+unchanged, then you can just fail to mention any rules in that chain,
+and it will be left untouched. (If you want to delete a chain you have
+to explicitly request that.)
+
+Thus, [the proposed PR] keeps track of which Service and EndpointSlice
+objects have changed since the last `iptables-restore`, and only
+outputs the Service-specific chains for a Service (the `KUBE-SVC-*`,
+`KUBE-SVL-*`, `KUBE-EXT-*`, `KUBE-FW-*`, and `KUBE-SEP-*` chains) if
+the Service or its EndpointSlice(s) changed (or were deleted) since the last
+`iptables-restore`. (The `KUBE-SERVICES`, `KUBE-EXTERNAL-SERVICES`,
+and `KUBE-NODEPORTS` chains are written in their entirety on every
+sync.)
+
```

Comment:
is there is a situation where one chain update will put the system in an unpredictable state (even for a very short while)? e.g., chain `KUBE-SVC-*` written out to iptables but the rest of chains are yet to be written?

if so (and i fully understand how much effort it entails, I just want to make sure that we have covered all options) wouldn't it be safer to re arrange rules in a way that one service update can - only - update a service specific chain per table? that way we don't relay on how clever the code is but rather on the structure of rules themselves

Reply:
I don't think we can do that without multiple writes (or some nasty pre-allocation scheme).  At the root of the tree is the KUBE_SERVICES chain which has a list of conditions that mean "this packet is service X".  For every service add/remove we need to add to that chain.

Reply:
> e.g., chain `KUBE-SVC-*` written out to iptables but the rest of chains are yet to be written?

Each `iptables-restore` is applied atomically. So the only way things would get out of sync would be if we specifically wrote out a set of out-of-sync rules.

> wouldn't it be safer to re arrange rules in a way that one service update can - only - update a service specific chain per table? that way we don't relay on how clever the code is but rather on the structure of rules themselves

Using `iptables-restore` to load the rules essentially requires that there be a chain like `KUBE-SERVICES` that contains all the services. (And _not_ using `iptables-restore` would absolutely destroy our performance.)